### PR TITLE
[topo_m0] Add new topo for mgmt tor router

### PR DESCRIPTION
**Pull request by yaqiangz**
### Description of PR

Summary:
Fixes [#5521](https://github.com/sonic-net/sonic-mgmt/issues/5521)

### Type of change

- [ ] Bug fix
- [x] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### Back port request
- [ ] 201911
- [x] 202012

#### What is the motivation for this PR?
For mgmt tor router, both testing methodology and physical wiring pattern are quite different with existing T0/T1/ topo. Adding this topo can let us filter unrelated test cases from mgmt network.

#### How did you do it?
* Add a new topo (inherit from t0-52)
* Enhance topo config and topo_facts for including interfaces
* Enhance minigraph templates to support generate minigraph for this new topo

#### How did you verify/test it?
Generate minigraph and deploy to DUT.
Use ./testbed-cli.sh deploy-mg to gen&deploy minigraph with new topo m0 successfully,
All the related configuration are correct pushed to DUT after deploy.

#### Any platform specific information?
N/A

### CODE REVIEW

File: ansible/library/announce_routes.py [hunk 1]
```diff
@@ -73,12 +73,12 @@ def wait_for_http(host_ip, http_port, timeout=10):
...
 
 def get_topo_type(topo_name):
-    pattern = re.compile(r'^(t0-mclag|t0|t1|ptf|fullmesh|dualtor|t2|mgmttor)')
+    pattern = re.compile(r'^(t0-mclag|t0|t1|ptf|fullmesh|dualtor|t2|mgmttor|m0)')
     match = pattern.match(topo_name)
     if not match:
         return "unsupported"
     topo_type = match.group()
-    if topo_type in ['dualtor', 'mgmttor']:
+    if topo_type in ['dualtor', 'mgmttor', 'm0']:
```
Comment (Blueve) [inline ansible/library/announce_routes.py:81]: We don't expect m0 to be a t0 topo

File: tests/common/testbed.py [hunk 2]
```diff
@@ -260,13 +260,13 @@ def _generate_sai_ptf_topo(self, tb_dict):
...
     def get_testbed_type(self, topo_name):
-        pattern = re.compile(r'^(wan|t0|t1|ptf|fullmesh|dualtor|t2|tgen|mgmttor)')
+        pattern = re.compile(r'^(wan|t0|t1|ptf|fullmesh|dualtor|t2|tgen|mgmttor|m0)')
         match = pattern.match(topo_name)
         if match == None:
             logger.warning("Unsupported testbed type - {}".format(topo_name))
             return "unsupported"
         tb_type = match.group()
-        if tb_type in ['mgmttor', 'dualtor']:
+        if tb_type in ['mgmttor', 'dualtor', 'm0']:
```
Comment (Blueve) [inline tests/common/testbed.py:269]: Same, we don't expect m0 to be t0

File: ansible/roles/eos/templates/m0-leaf.j2 [hunk 3]
```diff
@@ -0,0 +1 @@
+t0-leaf.j2
```
Comment (Blueve) [inline ansible/roles/eos/templates/m0-leaf.j2:1]: What is the different btw this one and https://github.com/sonic-net/sonic-mgmt/blob/7526b1e9c6c3e7f10e8ff10d038287fe67edcb1c/ansible/roles/eos/templates/t1-lag-spine.j2 

For M1, we can use t0-leaf for eos config generation for sure. But for Mx, are we referring the same>

File: ansible/vars/topo_m0.yml [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+        - 50
+      vm_offset: 4
+    ARISTA06MX:
+      vlans:
+        - 51
+      vm_offset: 5
+  DUT:
+    vlan_configs:
+      default_vlan_config: one_vlan_a
+      one_vlan_a:
```
Comment (Blueve) [inline ansible/vars/topo_m0.yml:77]: I think we can keep `one_vlan_a` here only.

@wangxin Are `two_vlan_a` / `three_vlan_a`/... required for each topo? Have them might be useful for some general usage... but I have no idea on which test cases are relay on these topos.

File: ansible/vars/topo_m0.yml [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+        ipv6: fc00::7e/126
+    bp_interface:
+      ipv4: 10.10.246.32/24
+      ipv6: fc0a::41/64
+
+  ARISTA05MX:
+    properties:
+    - common
+    bgp:
+      asn: 64600
```
Comment (Blueve) [inline ansible/vars/topo_m0.yml:233]: MX should use a different ASN as it is downstream of M0. Please check above comments.

File: ansible/vars/topo_m0.yml [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+configuration_properties:
+  common:
+    dut_asn: 65100
+    dut_type: MgmtToRRouter
+    podset_number: 30
+    tor_number: 16
+    tor_subnet_number: 2
+    max_tor_subnet_number: 16
+    tor_subnet_size: 128
+    spine_asn: 65534
```
Comment (Blueve) [inline ansible/vars/topo_m0.yml:132]: Let's understand where we use these three values: spine_asn, leaf_asn, tor_asn
M0 is more like a leaf here, and all M1 suppose to use spine_asn and Mx to use tor_asn

**Review by Blueve — CHANGES_REQUESTED** [review]

File: ansible/library/announce_routes.py (hunk 1 above)
Comment (yaqiangz) [reply ansible/library/announce_routes.py]: Fixed

File: tests/common/testbed.py (hunk 2 above)
Comment (yaqiangz) [reply tests/common/testbed.py]: Fixed

File: ansible/roles/eos/templates/m0-leaf.j2 (hunk 3 above)
Comment (yaqiangz) [reply ansible/roles/eos/templates/m0-leaf.j2:1]: Fixed, add a new template (m0-tor.j2) to generate config for Mx.

File: ansible/vars/topo_m0.yml (hunk 5 above)
Comment (yaqiangz) [reply ansible/vars/topo_m0.yml]: Fixed

File: ansible/vars/topo_m0.yml (hunk 6 above)
Comment (yaqiangz) [reply ansible/vars/topo_m0.yml]: Fixed

File: ansible/vars/topo_m0.yml (hunk 4 above)
Comment (yaqiangz) [reply ansible/vars/topo_m0.yml:77]: After discussing with @Blueve and @wangxin , keep one_vlan_a here only.

**Review by Blueve — APPROVED** [review]

**Comment by wangxin** [comment]
This change is included in 202012 branch by separate PR #6285

**Pull request merged**